network/p2p/middleware.go

```diff
@@ -152,6 +154,15 @@ func NewMiddleware(
 		opt(mw)
 	}
 
+	mw.ComponentManager = component.NewComponentManagerBuilder().OnStart(func(ctx context.Context) error {
+		// TODO: refactor to avoid storing ctx altogether
+		mw.ctx = ctx
+		return mw.start(ctx)
+	}).AddWorker(func(ctx irrecoverable.SignalerContext) {
+		<-ctx.Done()
+		mw.stop()
```

Comment:
This seems like something most/all implementors would need. Should that be more explicitly exposed in the API?

Reply:
not necessarily the stop method, but the context driven shutdown

Reply:
This is an anti-pattern, and I'm only doing it this way here because it would be too much work at the moment to completely refactor Middleware.

The idea with using contexts, is that there should never be any need for a "context driven shutdown". Any goroutines / processes which are started when the component is initiated should be passed in the given context to begin with, so they can check themselves whether it is time to shut down. They should not need to be told by a central authority when it is time to stop. This is the philosophy behind using contexts.

Reply:
There a book called Concurrency in Go that I read a while ago that talked a lot about this, but I don't remember the exact section that talks about this.

In this specific scenario, ideally any shutdown logic should be setup inside the call to `mw.start`, since that is already being passed in a context.

Reply:
fair enough. perhaps add a comment here to notify anyone that's using this as an example implementation